PIVX-Labs/MyPIVXWallet
# I18n tools

## Abstract

Add language tools to speed up localization development. 
The new proposed workflows will be as follows:
### Developers adding a new locale string
- Add the string in `template/translation.toml` and set it to the english value
- Run the script `lang-tools/translation_tamer.py`, this will add an empty string to all the template files.
### Translators
- Open the translation file
- Find the empty strings and translate accordingly
### Developers pushing a new release
~~In order to prevent duplicates, the `merge.py` script was introduced. Some localizations, have a lot of duplicate values, for example in en-US and en-GB (yes, we don't have them yet) it's mostly the same translation, save for a few strings ('color' vs 'colour'). The merge script takes care of this, removing the duplicates and placing them in another folder.
As of right now, only `pt-PT` and `pt-BR` require merging. Because of this, before each release it's recommended to run `lang-tools/merge.py pt-br/translation.toml pt-pt/translation.toml`.~~
add_string now automatically merges unless `--no-merge` is passed

## Additional i18n improvements/changes
- Added a language negotiation. Some people have different languages set in their browser/OS, for example they might be russians, but they also speak good french, but don't understand english at all. A correctly configured `navigator.languages` would return something along the lines of `['ru-RU', 'fr-FR']`. Previously MPW would use the english translation, despite having french available. Another example would be of a spanish user, having their `navigator.languages` return `['es-ES']`. While it is true that we don't have support for their specific locale, we have 'es-MX', which should be good enough. MPW should correctly behave in both of these examples
- Changed the translation files from `.js` to `.toml`. `JSON` was briefly considered, but dismissed because it doesn't support comments. This format should be easy to figure out for translators not familiar with programming, as it doesn't require any indentation, it's just `key = "translation"`. Under the hood, webpack converts the toml files to JSON, so it doesn't require any external parsing from the client
- Added auto language, which detects the language based on OS/browser preferences and set it as default

## Testing
- Check that all the strings are correctly translated in your locale
- Check that the auto language gives a sensible language based on your device settings

## Additional notes
Some languages were missing a few strings, specifically:
```
1 ./es-mx/translation.toml
1 ./fr/translation.toml
1 ./uwu/translation.toml
2 ./ph/translation.toml
6 ./de/translation.toml
10 ./it/translation.toml
```
Where the number represents the count of missing strings
